Allow clients to configure Kryo's class resolver and friends

CMoH opened the pull request:
Kryo can be configured to use a custom class resolver, but the Chill library hides the option from client code. Add appropriate constructor parameters to Chill's KryoBase subclass.

I needed such a custom class resolver to allow registration of custom serializers for classes belonging to dynamically loaded jars (plugins)

CMoH committed:
Allow clients to configure Kryo's class resolver and friends

Kryo can be configured to use a custom class resolver, but the Chill
library hides the option from client code. Add appropriate constructor
parameters to Chill's KryoBase subclass.

johnynek commented:
Thanks for the PR!

The intended way to customize your Kryo instance is with a KryoInstantiator:

https://github.com/twitter/chill/blob/develop/chill-java/src/main/java/com/twitter/chill/KryoInstantiator.java

Code the needs a Kryo instance should take that or a KryoPool and then you can instantiate the Kryo however you choose.

I hope that is a satisfactory approach for you. If not, I would recommend subclassing in your own code since adding more configuration to KryoBase is not the direction I think we should go (indeed the opposite, is rather remove that class entirely and move the logic to KryoInstantiators and IKryoRegistrars.

CMoH commented:
Thanks for taking the time to review this, @johnynek 

However, I can only find a single line of code that sets `Kryo.classResolver`, and that happens [here in Kryo's constructor](https://github.com/EsotericSoftware/kryo/blob/master/src/com/esotericsoftware/kryo/Kryo.java#L175). This constructor seems to be the only way of customizing this policy, and it makes sense from the perspective of Kryo devs, since the constructor also adds serializers to it later on.

`KryoBase` adds some Scala-related functions to the Java-oriented Kryo (like the object serializer etc). Should I copy-paste the KryoBase class into my code for now?

johnynek commented:
Looks like you are right that Kryo is missing the arguments to `set` these items after allocation.

Have you thought about sending a PR to kryo to add the set methods you need? They are pretty comfortable with mutable APIs.

I'd really like to move away from configuration with subclassing some massive base class. What I'd like is a compositional way to build these things up (which was the point of IKryoRegistrar and KryoInstantiator). I see the issue here is that `KryoBase` does include some of this.

What about this:

Can we add a default constructor (different from default arguments) so the previous zero arg constructor still exists? I think that will keep binary compatibility (and pass the MiMa check). If we can do that, I will merge the PR.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/twitter/chill?pullRequest=349) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/twitter/chill?pullRequest=349) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/twitter/chill?pullRequest=349) it.</sub>

CMoH committed:
Separate no-args constructor for binary compatibility

Replace default-args constructor with a pair of no-args constructor and
another that requires explicit arguments, thus preserving binary
compatibility

CMoH commented:
@johnynek - I noticed your comment today and reopened the issue. I've made the changes you requests, and I'm attempting to solve the merge conflicts now.

I will try adding the mutating methods you suggested on Kryo as well.

johnynek commented:
@CMoH I'm about to do a release. Can you sign the CLA and I'll merge this and release?

regadas commented on chill-scala/src/main/scala/com/twitter/chill/KryoBase.scala:
```diff
@@ -20,18 +20,29 @@ import com.esotericsoftware.kryo.KryoException
+class KryoBase(classResolver: ClassResolver, referenceResolver: ReferenceResolver)
+    extends Kryo(classResolver, referenceResolver) {
+  // no-args constructor instead of constructor with default args for binary compatibility
+  def this() {
```
```suggestion
  def this() = {
```

I think we should avoid procedure syntax. Also 🎨 we can remove the outer `{}`

johnynek replied:
is this procedure syntax? It doesn't return Unit, it is making a new instance of the class.

johnynek replied:
yeah, I can't get an alternate syntax to work... for an example at the REPL.

regadas replied:
yes right! it's not really a procedure, but i think that kinda of syntax should be avoid whenever possible. in this case only works because it's `this` otherwise it would return  `Unit`. regarding your error that's interesting I tried in 2.11 and 2.12 and had no issues.

```scala
[info] Starting scala interpreter...
Welcome to Scala 2.11.12 (OpenJDK 64-Bit Server VM, Java 1.8.0_222).
Type in expressions for evaluation. Or try :help.

scala> import com.twitter.chill._
import com.twitter.chill._

scala> import com.esotericsoftware.kryo.util.DefaultClassResolver
import com.esotericsoftware.kryo.util.DefaultClassResolver

scala> import com.esotericsoftware.kryo.util.MapReferenceResolver
import com.esotericsoftware.kryo.util.MapReferenceResolver

scala> new KryoBase()
res0: com.twitter.chill.KryoBase = com.twitter.chill.KryoBase@23c54aa6

scala> new KryoBase(new DefaultClassResolver, new MapReferenceResolver)
res1: com.twitter.chill.KryoBase = com.twitter.chill.KryoBase@68c54aa4
```

```scala
[info] Starting scala interpreter...
Welcome to Scala 2.12.8 (OpenJDK 64-Bit Server VM, Java 1.8.0_222).
Type in expressions for evaluation. Or try :help.

scala> import com.twitter.chill._
import com.twitter.chill._

scala> import com.esotericsoftware.kryo.util.DefaultClassResolver
import com.esotericsoftware.kryo.util.DefaultClassResolver

scala> import com.esotericsoftware.kryo.util.MapReferenceResolver
import com.esotericsoftware.kryo.util.MapReferenceResolver

scala> new KryoBase(new DefaultClassResolver, new MapReferenceResolver)
res0: com.twitter.chill.KryoBase = com.twitter.chill.KryoBase@58a12ac4

scala> new KryoBase()
res1: com.twitter.chill.KryoBase = com.twitter.chill.KryoBase@5b6e7715
```

johnynek replied:
Sorry, I didn't mean this code didn't work, I mean `def this() = ` didn't work.

I agree in scala every class should have 1 constructor, but binary compatibility is very important, so this is worth breaking the rule.

regadas replied:
🤔 but those REPL where ran with `def this() =` I'm a bit puzzled now 😄

johnynek replied:
Huh. Well if it works great. I prefer your syntax. I must have messed something up.

CMoH replied:
Glad to hear you solved it, and that it works fine!

johnynek replied:
@CMoH we still need the CLA signed, just a nagging reminder (also, I think @regadas would like you to update the code to use the `=`).

johnynek commented:
since this change is binary compatible, we can pick it up in the next version.

I want to go ahead and get the latest published as @regadas is waiting.

johnynek commented:
@CMoH I'm still willing to merge this if you will click the CLA.

CMoH commented:
Hi @johnynek - thanks for the reminder, but I have signed the CLA as soon as I noticed your post 28 days ago. Since nothing seemed to happen on the spot, I decided to let the systems sync (presumably).

Since nothing changed, I tried to fiddle with it tonight, but something seems to be wrong with the CLA signing software. See here: https://www.youtube.com/watch?v=AZqgXVt5ppY


Furthermore, even after following their steps, the CLA status is still pending:

![image](https://user-images.githubusercontent.com/663132/70945298-8db94c80-205d-11ea-8f81-3681dbb991ce.png)

Not sure how to go further - file a bug report with cla-assistant.io perhaps?

grimreaper commented:
@CMoH I'm an administrator for Twitter's github and directly added your username as having signed. If you wish please do file a bug with cla-assistant.io (https://github.com/cla-assistant/cla-assistant/issues/new). Also check adblockers and privacy settings (for example changing referer or cookies). 

@johnynek this is safe to merge now if you're having trouble let me know which merge method to use.

johnynek commented:
I cannot press the merge button while that CLA bot is yellow.

I guess I have merge rights but not admin rights and it requires more privileges than I have.

Can you merge it?

grimreaper commented:
done. If this remains an issue for other PRs tag me and I'll look into it.

johnynek commented:
Thanks!

On Tue, Dec 17, 2019 at 7:30 AM Eitan Adler <notifications@github.com>
wrote:

> done. If this remains an issue for other PRs tag me and I'll look into it.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/twitter/chill/pull/349?email_source=notifications&email_token=AAAQS5W6V6SENKPY4MGOHGLQZEEEHA5CNFSM4JJAZER2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEHDKIFY#issuecomment-566666263>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAQS5X5G6A43CGSPTI47VTQZEEEHANCNFSM4JJAZERQ>
> .
>


-- 
P. Oscar Boykin, Ph.D. | http://twitter.com/posco | http://pobox.com/~boykin

CMoH commented:
Thank you all!
